spec/models/user/academics/term_registration_spec.rb

```diff
@@ -0,0 +1,299 @@
+require 'spec_helper'
+
+RSpec.describe User::Academics::TermRegistration, type: :model do
+  let(:tuition_calculated) do
+    User::Academics::StudentAttribute.new(tuition_calculated: true)
+  end
+
+  let(:registered) {
+    User::Academics::StudentAttribute.new(registered: true)
+  }
+
+  let(:student_attributes) {
+    instance_double("User::Academics::StudentAttributes")
+  }
+
+  let(:twenty_percent_cnp_exception) {
+    User::Academics::StudentAttribute.new(twenty_percent_cnp_exception: true)
+  }
+
+  let(:term) {
+    instance_double("User::Academics::Term").tap do |mock|
+      allow(mock).to receive(:term_id).and_return("BOGUS-TERM-ID")
+    end
+  }
+
+  let(:registrations) { double("User::Academics::Registrations") }
+
+  let(:undergrad_registration) {
+    instance_double("User::Academics::Registration").tap do |mock|
+      allow(mock).to receive(:graduate?).and_return(false)
+      allow(mock).to receive(:law?).and_return(false)
+      allow(mock).to receive(:undergraduate?).and_return(true)
+    end
+  }
+
+  let(:graduate_registration) {
+    instance_double("User::Academics::Registration").tap do |mock|
+      allow(mock).to receive(:graduate?).and_return(true)
+      allow(mock).to receive(:law?).and_return(false)
+      allow(mock).to receive(:undergraduate?).and_return(false)
+    end
+  }
+
+  let(:law_registration) {
+    instance_double("User::Academics::Registration").tap do |mock|
+      allow(mock).to receive(:graduate?).and_return(false)
+      allow(mock).to receive(:law?).and_return(true)
+      allow(mock).to receive(:undergraduate?).and_return(false)
+    end
+  }
+
+  describe "for a summer term" do
```

Comment:
@developish - I want to discuss your style of Rspec here in more depth. I'm trying to model a way of developing the specs so that they're isolated units, with even the methods being isolated. We've applied this kind of general "in this scenario the model should do this, in that scenario the model should do that" type of thing, which integrates all the methods together, even involving other external classes/models executing, and it seems to end up making for extremely complex test harnesses that to the initial developer seem easy to maintain, but really they grow to become extremely complex and hard to cognitively hold in ones mind when analyzing and modifying.

An example of this approach can be seen in [MyAcademics::Semesters spec](https://github.com/sis-berkeley-edu/calcentral/blob/v103.2/spec/models/my_academics/semesters_spec.rb).

Ping me when you see this so we can discuss on Slack further.

Reply:
@developish explained that his approach here was to ensure that the high level behavioral specifications obtained from BSA's / Designers is documented in this top-level model spec, acting as a form of integration test. He has plans to migrate some testing to a lower-model level to make it more maintainable.

I explained how my approach has been to lobby for the isolation of tests, even on the model level, and allow local development environment or staging (dev, test, QA) servers be used for integration testing. He disagreed, pointing out that we'd lose the opportunity to test for scenarios that the code handles that require much time and many personnel to setup, (e.g. testing summer scenarios when we're not currently in summer). I agreed with this point.

The balanced approach to testing that we both agreed on is that

1. With the approach to implementing simpler interfaces, the mocking in tests should require much fewer lines of code for the mocked data or mocking. 
2. We don't want to sacrifice explaining how things work on a higher level, in a way that communicates the specifications to future developers
3. When needed, we should migrate complex functionality and their specifications to lower level models, thus abstracting the away the need for that coverage in the higher level integration spec